addon/addon-test-support/@ember/test-helpers/-internal/-owner-mixin-imports.ts

```diff
@@ -0,0 +1,78 @@
+// These utter shenanigans allow this to work regardless of what types you are
+// importing from. The idea is: you supply multiple imports, each of which is
+// mutually exclusive with the others in terms of valid imports, then ignore the
+// imports so they type check, then "resolve" between them based on which one is
+// actually defined. 😭
+//
+// TODO: at the next major version (4.0), when we drop support for Ember < 5.1
+// or some later version and can therefore either depend *somewhat* more on
+// these import locations *or* can drop them entirely (depending on the path
+// taken in terms of the `Owner` stack), switch to the normal imports:
+// ```
+// import {
+//   ContainerProxyMixin,
+//   RegistryProxyMixin,
+// } from '@ember/-internals/runtime';
+// ```
+//
+// For now, we are stuck with this utterly horrifying hack to get the types for
+// these reliably.
+
+// This pair of types lets us check whether the imported type is set. We can
+// then use that to pick between the imported types. The first type produces
+// `any` when the type is otherwise not defined at all, which means we cannot
+// use it directly because `any` will fall through, well, basically *any* check
+// we might come up with... other than checking whether `any extends` it. The
+// second uses that to get back a `never` so we can then "pick" between the
+// imported types.
+type AnyIfNeverElseT<T> = T extends never ? never : T;
+type IsNever<T> = any extends AnyIfNeverElseT<T> ? never : T;
+
+// This mapped type does the "picking". You can pass it any number of types,
+// where only one should be available, and it will "resolve" that type.
+// Constraint: the `Array` type must include mutually exclusive types, such that
+// it should be impossible to ever get back more than one; otherwise, you will
+// end up with a union type of the two. That is very unlikely to be what you
+// want!
+type NonNever<Types extends Array<unknown>> = {
+  [Index in keyof Types]: IsNever<Types[Index]>;
+}[number];
+
+// Imports from `@types`
+// @ts-ignore
+import type CPM_DTS from '@ember/engine/-private/container-proxy-mixin.d.ts';
```

Comment:
Can you ELI5 this?

Reply:
Roughly, it imports from all the different paths where this type can exist, using `ts-ignore` on all of them to make the ones which are not present on any given Ember types version still pass the type checking, and then go through these type shenanigans so that it only resolves the one that actually does exist on the current version. The net result is that whether you are using the Definitely Typed types, the 4:8 preview types, the 4.12 preview types, or the stable types, one of these will be available and the others will not, and these exports will correctly resolve only that one. If you don’t do the type level shenanigans, you end up mixing `any` in (since that is the type of a `ts-ignore`d import which has no type definition at that module path), which is definitely not what we want.